x-pack/elastic-agent/pkg/agent/cmd/reexec_windows.go

```diff
@@ -0,0 +1,76 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+// +build windows
+
+package cmd
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/spf13/cobra"
+	"golang.org/x/sys/windows/svc"
+	"golang.org/x/sys/windows/svc/mgr"
+
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/errors"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/cli"
+)
+
+func newReExecWindowsCommand(flags *globalFlags, _ []string, streams *cli.IOStreams) *cobra.Command {
+	cmd := &cobra.Command{
+		Hidden: true,
+		Use:    "reexec_windows <service_name> <pid>",
+		Short:  "ReExec the windows service",
+		Long:   "This waits for the windows service to stop then restarts it to allow self-upgrading.",
+		Args:   cobra.ExactArgs(2),
+		Run: func(c *cobra.Command, args []string) {
+			serviceName := args[0]
+			servicePid, err := strconv.Atoi(args[1])
+			if err != nil {
+				fmt.Fprintf(streams.Err, "%v\n", err)
+				os.Exit(1)
+			}
+			err = reExec(serviceName, servicePid)
+			if err != nil {
+				fmt.Fprintf(streams.Err, "%v\n", err)
+				os.Exit(1)
+			}
+		},
+	}
+
+	return cmd
+}
+
+func reExec(serviceName string, servicePid int) error {
+	manager, err := mgr.Connect()
+	if err != nil {
+		return errors.New(err, "failed to connect to service manager")
+	}
+	service, err := manager.OpenService(serviceName)
+	if err != nil {
+		return errors.New(err, "failed to open service")
+	}
+	for {
+		status, err := service.Query()
+		if err != nil {
+			return errors.New(err, "failed to query service")
+		}
+		if status.State == svc.Stopped {
+			err = service.Start()
```

Comment:
do we need to handle case when service is set to auto-restart?

Reply:
It does handle this case, by checking the `service.ProcessId` with the previous `servicePid`. When the command is called the previous pid is sent in the arguments to the command. Inside the loop it checks if the service is stopped (if so it starts it, and exists) if its not stopped it checks the current `service.ProcessId` against the previous pid and if it is different then it means it was already restarted so we also exit.